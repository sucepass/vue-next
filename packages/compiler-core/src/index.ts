export { baseCompile } from './compile'

// Also expose lower level APIs & types
export {
  CompilerOptions,
  ParserOptions,
  TransformOptions,
  CodegenOptions
} from './options'
export { baseParse, TextModes } from './parse'
export {
  transform,
  createStructuralDirectiveTransform,
  TransformContext,
  NodeTransform,
  StructuralDirectiveTransform,
  DirectiveTransform
} from './transform'
export { generate, CodegenContext, CodegenResult } from './codegen'
export {
  ErrorCodes,
  CoreCompilerError,
  CompilerError,
  createCompilerError
} from './errors'
export * from './ast'
export * from './utils'
export { registerRuntimeHelpers } from './runtimeHelpers'

// expose transforms so higher-order compilers can import and extend them
export { transformModel } from './transforms/vModel'
export { transformOn } from './transforms/vOn'

// utility, but need to rewrite typing to avoid dts relying on @vue/shared
import { generateCodeFrame as _genCodeFrame } from '@vue/shared'
const generateCodeFrame = _genCodeFrame as (
  source: string,
  start?: number,
  end?: number
) => string
export { generateCodeFrame }
