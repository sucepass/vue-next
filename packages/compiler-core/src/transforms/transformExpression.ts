// - Parse expressions in templates into compound expressions so that each
//   identifier gets more accurate source-map locations.
//
// - Prefix identifiers with `_ctx.` so that they are accessed from the render
//   context
//
// - This transform is only applied in non-browser builds because it relies on
//   an additional JavaScript parser. In the browser, there is no source-map
//   support and the code is wrapped in `with (this) { ... }`.

import { parseScript } from 'meriyah'
import { walk } from 'estree-walker'
import { NodeTransform, TransformContext } from '../transform'
import { NodeTypes, createExpression, ExpressionNode } from '../ast'
import { Node, Function, Identifier, Property } from 'estree'
import { advancePositionWithClone } from '../utils'
export const transformExpression: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.EXPRESSION && !node.isStatic) {
    processExpression(node, context)
  } else if (node.type === NodeTypes.ELEMENT) {
    // handle directives on element
    for (let i = 0; i < node.props.length; i++) {
      const prop = node.props[i]
      if (prop.type === NodeTypes.DIRECTIVE) {
        if (prop.exp) {
          processExpression(prop.exp, context)
        }
        if (prop.arg && !prop.arg.isStatic) {
          if (prop.name === 'class') {
            // TODO special expression optimization for classes
            processExpression(prop.arg, context)
          } else {
            processExpression(prop.arg, context)
          }
        }
      }
    }
  }
}

const simpleIdRE = /^[a-zA-Z$_][\w$]*$/

const isFunction = (node: Node): node is Function =>
  /Function(Expression|Declaration)$/.test(node.type)

const isPropertyKey = (node: Node, parent: Node) =>
  parent.type === 'Property' && parent.key === node && !parent.computed

// cache node requires
let _parseScript: typeof parseScript
let _walk: typeof walk

interface PrefixMeta {
  prefix?: string
  start: number
  end: number
}

// Important: since this function uses Node.js only dependencies, it should
// always be used with a leading !__BROWSER__ check so that it can be
// tree-shaken from the browser build.
export function processExpression(
  node: ExpressionNode,
  context: TransformContext
) {
  if (!context.prefixIdentifiers) {
    return
  }
  // lazy require dependencies so that they don't end up in rollup's dep graph
  // and thus can be tree-shaken in browser builds.
  const parseScript =
    _parseScript || (_parseScript = require('meriyah').parseScript)
  const walk = _walk || (_walk = require('estree-walker').walk)

  // fast path if expression is a simple identifier.
  if (simpleIdRE.test(node.content)) {
    if (!context.identifiers[node.content]) {
      node.content = `_ctx.${node.content}`
    }
    return
  }

  let ast
  try {
    ast = parseScript(`(${node.content})`, { ranges: true }) as any
  } catch (e) {
    context.onError(e)
    return
  }

  const ids: (Identifier & PrefixMeta)[] = []
  const knownIds = Object.create(context.identifiers)

  // walk the AST and look for identifiers that need to be prefixed with `_ctx.`.
  walk(ast, {
    enter(node: Node & PrefixMeta, parent) {
      if (node.type === 'Identifier') {
        if (!ids.includes(node)) {
          if (!knownIds[node.name] && shouldPrefix(node, parent)) {
            if (
              isPropertyKey(node, parent) &&
              (parent as Property).value === node
            ) {
              // property shorthand like { foo }, we need to add the key since we
              // rewrite the value
              node.prefix = `${node.name}: `
            }
            node.name = `_ctx.${node.name}`
            ids.push(node)
          } else if (!isPropertyKey(node, parent)) {
            // also generate sub-expressioms for other identifiers for better
            // source map support. (except for property keys which are static)
            ids.push(node)
          }
        }
      } else if (isFunction(node)) {
        // walk function expressions and add its arguments to known identifiers
        // so that we don't prefix them
        node.params.forEach(p =>
          walk(p, {
            enter(child) {
              if (child.type === 'Identifier') {
                knownIds[child.name] = true
                ;(
                  (node as any)._scopeIds ||
                  ((node as any)._scopeIds = new Set())
                ).add(child.name)
              }
            }
          })
        )
      }
    },
    leave(node: any) {
      if (node._scopeIds) {
        node._scopeIds.forEach((id: string) => {
          delete knownIds[id]
        })
      }
    }
  })

  // We break up the coumpound expression into an array of strings and sub
  // expressions (for identifiers that have been prefixed). In codegen, if
  // an ExpressionNode has the `.children` property, it will be used instead of
  // `.content`.
  const full = node.content
  const children: ExpressionNode['children'] = []
  ids.sort((a, b) => a.start - b.start)
  ids.forEach((id, i) => {
    const last = ids[i - 1] as any
    const leadingText = full.slice(last ? last.end - 1 : 0, id.start - 1)
    if (leadingText.length || id.prefix) {
      children.push(leadingText + (id.prefix || ``))
    }
    const source = full.slice(id.start - 1, id.end - 1)
    children.push(
      createExpression(id.name, false, {
        source,
        start: advancePositionWithClone(node.loc.start, source, id.start + 2),
        end: advancePositionWithClone(node.loc.start, source, id.end + 2)
      })
    )
    if (i === ids.length - 1 && id.end - 1 < full.length) {
      children.push(full.slice(id.end - 1))
    }
  })

  if (children.length) {
    // mark it empty so that it's more noticeable in case another transform or
    // codegen forget to handle `.children` first.
    node.content = __DEV__ ? `[[REMOVED]]` : ``
    node.children = children
  }
}

const globals = new Set(
  (
    'Infinity,undefined,NaN,isFinite,isNaN,' +
    'parseFloat,parseInt,decodeURI,decodeURIComponent,encodeURI,encodeURIComponent,' +
    'Math,Number,Date,Array,Object,Boolean,String,RegExp,Map,Set,JSON,Intl,' +
    'require,' + // for webpack
    'arguments,'
  ) // parsed as identifier but is a special keyword...
    .split(',')
)

function shouldPrefix(identifier: Identifier, parent: Node) {
  if (
    !(
      isFunction(parent) &&
      // not id of a FunctionDeclaration
      ((parent as any).id === identifier ||
        // not a params of a function
        parent.params.includes(identifier))
    ) &&
    // not a key of Property
    !(
      isPropertyKey(identifier, parent) &&
      // shorthand keys should be prefixed
      !((parent as Property).value === identifier)
    ) &&
    // not a property of a MemberExpression
    !(
      parent.type === 'MemberExpression' &&
      parent.property === identifier &&
      !parent.computed
    ) &&
    // not in an Array destructure pattern
    !(parent.type === 'ArrayPattern') &&
    // skip globals + commonly used shorthands
    !globals.has(identifier.name)
  ) {
    return true
  }
}
