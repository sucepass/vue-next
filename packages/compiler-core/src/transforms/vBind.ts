import { DirectiveTransform } from '../transform'
import { createObjectProperty, createExpression } from '../ast'
import { createCompilerError, ErrorCodes } from '../errors'
import { camelize } from '@vue/shared'

// v-bind without arg is handled directly in ./element.ts due to it affecting
// codegen for the entire props object. This transform here is only for v-bind
// *with* args.
export const transformBind: DirectiveTransform = (
  { exp, arg, modifiers, loc },
  context
) => {
  if (!exp) {
    context.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
  }
  // .prop is no longer necessary due to new patch behavior
  // .sync is replced by v-model:arg
  if (modifiers.includes('camel')) {
    arg!.content = camelize(arg!.content)
  }
  return {
    props: createObjectProperty(
      arg!,
      exp || createExpression('', true, loc),
      loc
    ),
    needRuntime: false
  }
}
