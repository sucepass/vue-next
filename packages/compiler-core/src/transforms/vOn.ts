import { DirectiveTransform } from '../transform'
import {
  createObjectProperty,
  createSimpleExpression,
  ExpressionNode,
  NodeTypes,
  createCompoundExpression,
  SimpleExpressionNode
} from '../ast'
import { capitalize } from '@vue/shared'
import { createCompilerError, ErrorCodes } from '../errors'
import { processExpression } from './transformExpression'
import { isMemberExpression } from '../utils'

const fnExpRE = /^([\w$_]+|\([^)]*?\))\s*=>|^function(?:\s+[\w$]+)?\s*\(/

// v-on without arg is handled directly in ./element.ts due to it affecting
// codegen for the entire props object. This transform here is only for v-on
// *with* args.
export const transformOn: DirectiveTransform = (dir, node, context) => {
  const { loc, modifiers } = dir
  const arg = dir.arg!
  if (!dir.exp && !modifiers.length) {
    context.onError(createCompilerError(ErrorCodes.X_V_ON_NO_EXPRESSION, loc))
  }
  let eventName: ExpressionNode
  if (arg.type === NodeTypes.SIMPLE_EXPRESSION) {
    if (arg.isStatic) {
      eventName = createSimpleExpression(
        `on${capitalize(arg.content)}`,
        true,
        arg.loc
      )
    } else {
      eventName = createCompoundExpression([`"on" + (`, arg, `)`])
    }
  } else {
    // already a compound expression.
    eventName = arg
    eventName.children.unshift(`"on" + (`)
    eventName.children.push(`)`)
  }
  // TODO .once modifier handling since it is platform agnostic
  // other modifiers are handled in compiler-dom

  // handler processing
  if (dir.exp) {
    // exp is guaranteed to be a simple expression here because v-on w/ arg is
    // skipped by transformExpression as a special case.
    let exp: ExpressionNode = dir.exp as SimpleExpressionNode
    const isInlineStatement = !(
      isMemberExpression(exp.content) || fnExpRE.test(exp.content)
    )
    // process the expression since it's been skipped
    if (!__BROWSER__ && context.prefixIdentifiers) {
      context.addIdentifiers(`$event`)
      exp = processExpression(exp, context)
      context.removeIdentifiers(`$event`)
    }
    if (isInlineStatement) {
      // wrap inline statement in a function expression
      exp = createCompoundExpression([
        `$event => (`,
        ...(exp.type === NodeTypes.SIMPLE_EXPRESSION ? [exp] : exp.children),
        `)`
      ])
    }
    dir.exp = exp
  }

  return {
    props: [
      createObjectProperty(
        eventName,
        dir.exp || createSimpleExpression(`() => {}`, false, loc)
      )
    ],
    needRuntime: false
  }
}
