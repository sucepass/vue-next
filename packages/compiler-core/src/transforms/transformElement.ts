import { NodeTransform, TransformContext } from '../transform'
import {
  NodeTypes,
  ElementTypes,
  CallExpression,
  ObjectExpression,
  ElementNode,
  DirectiveNode,
  ExpressionNode,
  ArrayExpression,
  createCallExpression,
  createArrayExpression,
  createObjectProperty,
  createSimpleExpression,
  createObjectExpression,
  Property
} from '../ast'
import { PatchFlags, PatchFlagNames, isSymbol } from '@vue/shared'
import { createCompilerError, ErrorCodes } from '../errors'
import {
  CREATE_VNODE,
  APPLY_DIRECTIVES,
  RESOLVE_DIRECTIVE,
  RESOLVE_COMPONENT,
  MERGE_PROPS,
  TO_HANDLERS
} from '../runtimeHelpers'
import { getInnerRange, isVSlot, toValidAssetId } from '../utils'
import { buildSlots } from './vSlot'

// some directive transforms (e.g. v-model) may return a symbol for runtime
// import, which should be used instead of a resolveDirective call.
const directiveImportMap = new WeakMap<DirectiveNode, symbol>()

// generate a JavaScript AST for this element's codegen
export const transformElement: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ELEMENT) {
    if (
      node.tagType === ElementTypes.ELEMENT ||
      node.tagType === ElementTypes.COMPONENT ||
      // <template> with v-if or v-for are ignored during traversal.
      // <template> without v-slot should be treated as a normal element.
      (node.tagType === ElementTypes.TEMPLATE && !node.props.some(isVSlot))
    ) {
      // perform the work on exit, after all child expressions have been
      // processed and merged.
      return () => {
        const isComponent = node.tagType === ElementTypes.COMPONENT
        let hasProps = node.props.length > 0
        let patchFlag: number = 0
        let runtimeDirectives: DirectiveNode[] | undefined
        let dynamicPropNames: string[] | undefined

        if (isComponent) {
          context.helper(RESOLVE_COMPONENT)
          context.components.add(node.tag)
        }

        const args: CallExpression['arguments'] = [
          isComponent ? toValidAssetId(node.tag, `component`) : `"${node.tag}"`
        ]
        // props
        if (hasProps) {
          const propsBuildResult = buildProps(node, context)
          patchFlag = propsBuildResult.patchFlag
          dynamicPropNames = propsBuildResult.dynamicPropNames
          runtimeDirectives = propsBuildResult.directives
          if (!propsBuildResult.props) {
            hasProps = false
          } else {
            args.push(propsBuildResult.props)
          }
        }
        // children
        const hasChildren = node.children.length > 0
        if (hasChildren) {
          if (!hasProps) {
            args.push(`null`)
          }
          if (isComponent) {
            const { slots, hasDynamicSlots } = buildSlots(node, context)
            args.push(slots)
            if (hasDynamicSlots) {
              patchFlag |= PatchFlags.DYNAMIC_SLOTS
            }
          } else if (node.children.length === 1) {
            const child = node.children[0]
            const type = child.type
            const hasDynamicTextChild =
              type === NodeTypes.INTERPOLATION ||
              type === NodeTypes.COMPOUND_EXPRESSION
            if (hasDynamicTextChild) {
              patchFlag |= PatchFlags.TEXT
            }
            // pass directly if the only child is a text node
            // (plain / interpolation / expression)
            if (hasDynamicTextChild || type === NodeTypes.TEXT) {
              args.push(child)
            } else {
              args.push(node.children)
            }
          } else {
            args.push(node.children)
          }
        }
        // patchFlag & dynamicPropNames
        if (patchFlag !== 0) {
          if (!hasChildren) {
            if (!hasProps) {
              args.push(`null`)
            }
            args.push(`null`)
          }
          if (__DEV__) {
            const flagNames = Object.keys(PatchFlagNames)
              .map(Number)
              .filter(n => n > 0 && patchFlag & n)
              .map(n => PatchFlagNames[n])
              .join(`, `)
            args.push(patchFlag + ` /* ${flagNames} */`)
          } else {
            args.push(patchFlag + '')
          }
          if (dynamicPropNames && dynamicPropNames.length) {
            args.push(
              `[${dynamicPropNames.map(n => JSON.stringify(n)).join(`, `)}]`
            )
          }
        }

        const { loc } = node
        const vnode = createCallExpression(
          context.helper(CREATE_VNODE),
          args,
          loc
        )

        if (runtimeDirectives && runtimeDirectives.length) {
          node.codegenNode = createCallExpression(
            context.helper(APPLY_DIRECTIVES),
            [
              vnode,
              createArrayExpression(
                runtimeDirectives.map(dir => buildDirectiveArgs(dir, context)),
                loc
              )
            ],
            loc
          )
        } else {
          node.codegenNode = vnode
        }
      }
    }
  }
}

export type PropsExpression = ObjectExpression | CallExpression | ExpressionNode

export function buildProps(
  node: ElementNode,
  context: TransformContext,
  props: ElementNode['props'] = node.props
): {
  props: PropsExpression | undefined
  directives: DirectiveNode[]
  patchFlag: number
  dynamicPropNames: string[]
} {
  const elementLoc = node.loc
  const isComponent = node.tagType === ElementTypes.COMPONENT
  let properties: ObjectExpression['properties'] = []
  const mergeArgs: PropsExpression[] = []
  const runtimeDirectives: DirectiveNode[] = []

  // patchFlag analysis
  let patchFlag = 0
  let hasRef = false
  let hasClassBinding = false
  let hasStyleBinding = false
  let hasDynamicKeys = false
  const dynamicPropNames: string[] = []

  const analyzePatchFlag = ({ key, value }: Property) => {
    if (key.type === NodeTypes.SIMPLE_EXPRESSION && key.isStatic) {
      if (value.type !== NodeTypes.SIMPLE_EXPRESSION || !value.isStatic) {
        const name = key.content
        if (name === 'ref') {
          hasRef = true
        } else if (name === 'class') {
          hasClassBinding = true
        } else if (name === 'style') {
          hasStyleBinding = true
        } else if (name !== 'key') {
          dynamicPropNames.push(name)
        }
      }
    } else {
      hasDynamicKeys = true
    }
  }

  for (let i = 0; i < props.length; i++) {
    // static attribute
    const prop = props[i]
    if (prop.type === NodeTypes.ATTRIBUTE) {
      const { loc, name, value } = prop
      if (name === 'ref') {
        hasRef = true
      }
      properties.push(
        createObjectProperty(
          createSimpleExpression(
            name,
            true,
            getInnerRange(loc, 0, name.length)
          ),
          createSimpleExpression(
            value ? value.content : '',
            true,
            value ? value.loc : loc
          )
        )
      )
    } else {
      // directives
      const { name, arg, exp, loc } = prop

      // skip v-slot - it is handled by its dedicated transform.
      if (name === 'slot') {
        if (!isComponent) {
          context.onError(
            createCompilerError(ErrorCodes.X_V_SLOT_MISPLACED, loc)
          )
        }
        continue
      }

      // special case for v-bind and v-on with no argument
      const isBind = name === 'bind'
      const isOn = name === 'on'
      if (!arg && (isBind || isOn)) {
        hasDynamicKeys = true
        if (exp) {
          if (properties.length) {
            mergeArgs.push(
              createObjectExpression(dedupeProperties(properties), elementLoc)
            )
            properties = []
          }
          if (isBind) {
            mergeArgs.push(exp)
          } else {
            // v-on="obj" -> toHandlers(obj)
            mergeArgs.push({
              type: NodeTypes.JS_CALL_EXPRESSION,
              loc,
              callee: context.helper(TO_HANDLERS),
              arguments: [exp]
            })
          }
        } else {
          context.onError(
            createCompilerError(
              isBind
                ? ErrorCodes.X_V_BIND_NO_EXPRESSION
                : ErrorCodes.X_V_ON_NO_EXPRESSION,
              loc
            )
          )
        }
        continue
      }

      const directiveTransform = context.directiveTransforms[name]
      if (directiveTransform) {
        // has built-in directive transform.
        const { props, needRuntime } = directiveTransform(prop, node, context)
        props.forEach(analyzePatchFlag)
        properties.push(...props)
        if (needRuntime) {
          runtimeDirectives.push(prop)
          if (isSymbol(needRuntime)) {
            directiveImportMap.set(prop, needRuntime)
          }
        }
      } else {
        // no built-in transform, this is a user custom directive.
        runtimeDirectives.push(prop)
      }
    }
  }

  let propsExpression: PropsExpression | undefined = undefined

  // has v-bind="object" or v-on="object", wrap with mergeProps
  if (mergeArgs.length) {
    if (properties.length) {
      mergeArgs.push(
        createObjectExpression(dedupeProperties(properties), elementLoc)
      )
    }
    if (mergeArgs.length > 1) {
      propsExpression = createCallExpression(
        context.helper(MERGE_PROPS),
        mergeArgs,
        elementLoc
      )
    } else {
      // single v-bind with nothing else - no need for a mergeProps call
      propsExpression = mergeArgs[0]
    }
  } else if (properties.length) {
    propsExpression = createObjectExpression(
      dedupeProperties(properties),
      elementLoc
    )
  }

  // patchFlag analysis
  if (hasDynamicKeys) {
    patchFlag |= PatchFlags.FULL_PROPS
  } else {
    if (hasClassBinding) {
      patchFlag |= PatchFlags.CLASS
    }
    if (hasStyleBinding) {
      patchFlag |= PatchFlags.STYLE
    }
    if (dynamicPropNames.length) {
      patchFlag |= PatchFlags.PROPS
    }
  }
  if (patchFlag === 0 && (hasRef || runtimeDirectives.length > 0)) {
    patchFlag |= PatchFlags.NEED_PATCH
  }

  return {
    props: propsExpression,
    directives: runtimeDirectives,
    patchFlag,
    dynamicPropNames
  }
}

// Dedupe props in an object literal.
// Literal duplicated attributes would have been warned during the parse phase,
// however, it's possible to encounter duplicated `onXXX` handlers with different
// modifiers. We also need to merge static and dynamic class / style attributes.
// - onXXX handlers / style: merge into array
// - class: merge into single expression with concatenation
function dedupeProperties(properties: Property[]): Property[] {
  const knownProps: Record<string, Property> = {}
  const deduped: Property[] = []
  for (let i = 0; i < properties.length; i++) {
    const prop = properties[i]
    // dynamic keys are always allowed
    if (prop.key.type === NodeTypes.COMPOUND_EXPRESSION || !prop.key.isStatic) {
      deduped.push(prop)
      continue
    }
    const name = prop.key.content
    const existing = knownProps[name]
    if (existing) {
      if (
        name === 'style' ||
        name === 'class' ||
        name.startsWith('on') ||
        name.startsWith('vnode')
      ) {
        mergeAsArray(existing, prop)
      }
      // unexpected duplicate, should have emitted error during parse
    } else {
      knownProps[name] = prop
      deduped.push(prop)
    }
  }
  return deduped
}

function mergeAsArray(existing: Property, incoming: Property) {
  if (existing.value.type === NodeTypes.JS_ARRAY_EXPRESSION) {
    existing.value.elements.push(incoming.value)
  } else {
    existing.value = createArrayExpression(
      [existing.value, incoming.value],
      existing.loc
    )
  }
}

function buildDirectiveArgs(
  dir: DirectiveNode,
  context: TransformContext
): ArrayExpression {
  const dirArgs: ArrayExpression['elements'] = []
  const runtime = directiveImportMap.get(dir)
  if (runtime) {
    context.helper(runtime)
    dirArgs.push(context.helperString(runtime))
  } else {
    // inject statement for resolving directive
    context.helper(RESOLVE_DIRECTIVE)
    context.directives.add(dir.name)
    dirArgs.push(toValidAssetId(dir.name, `directive`))
  }
  const { loc } = dir
  if (dir.exp) dirArgs.push(dir.exp)
  if (dir.arg) {
    if (!dir.exp) {
      dirArgs.push(`void 0`)
    }
    dirArgs.push(dir.arg)
  }
  if (Object.keys(dir.modifiers).length) {
    if (!dir.arg) {
      if (!dir.exp) {
        dirArgs.push(`void 0`)
      }
      dirArgs.push(`void 0`)
    }
    dirArgs.push(
      createObjectExpression(
        dir.modifiers.map(modifier =>
          createObjectProperty(
            modifier,
            createSimpleExpression(`true`, false, loc)
          )
        ),
        loc
      )
    )
  }
  return createArrayExpression(dirArgs, dir.loc)
}
