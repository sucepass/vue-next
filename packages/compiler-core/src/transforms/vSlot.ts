import {
  ElementNode,
  ObjectExpression,
  createObjectExpression,
  NodeTypes,
  createCompoundExpression,
  createCallExpression,
  CompoundExpressionNode,
  CallExpression,
  createObjectProperty,
  createSimpleExpression,
  createFunctionExpression,
  DirectiveNode,
  ElementTypes,
  ExpressionNode,
  Property,
  ChildNode,
  SourceLocation
} from '../ast'
import { TransformContext } from '../transform'
import { buildProps } from './transformElement'
import { createCompilerError, ErrorCodes } from '../errors'
import { isSimpleIdentifier } from '../utils'
import { RENDER_SLOT } from '../runtimeConstants'
import { isString } from '@vue/shared'

const isVSlot = (p: ElementNode['props'][0]): p is DirectiveNode =>
  p.type === NodeTypes.DIRECTIVE && p.name === 'slot'

export function buildSlots(
  { props, children, loc }: ElementNode,
  context: TransformContext
): ObjectExpression {
  const slots: Property[] = []

  // 1. Check for default slot with slotProps on component itself.
  //    <Comp v-slot="{ prop }"/>
  const explicitDefaultSlot = props.find(isVSlot)
  if (explicitDefaultSlot) {
    const { arg, exp, loc } = explicitDefaultSlot
    if (arg) {
      context.onError(
        createCompilerError(ErrorCodes.X_NAMED_SLOT_ON_COMPONENT, loc)
      )
    }
    slots.push(buildSlot(`default`, exp, children, loc))
  }

  // 2. Iterate through children and check for template slots
  //    <template v-slot:foo="{ prop }">
  let hasTemplateSlots = false
  const seenSlotNames = new Set<string>()
  const nonSlotChildren: ChildNode[] = []
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    if (
      child.type === NodeTypes.ELEMENT &&
      child.tagType === ElementTypes.TEMPLATE
    ) {
      const { props, children, loc: nodeLoc } = child
      const slotDir = props.find(isVSlot)
      if (slotDir) {
        hasTemplateSlots = true
        const { arg: slotName, exp: slotProps, loc: dirLoc } = slotDir
        if (explicitDefaultSlot) {
          // already has on-component default slot - this is incorrect usage.
          context.onError(
            createCompilerError(ErrorCodes.X_MIXED_SLOT_USAGE, dirLoc)
          )
          break
        } else {
          // check duplicate slot names
          if (
            !slotName ||
            (slotName.type === NodeTypes.SIMPLE_EXPRESSION && slotName.isStatic)
          ) {
            const name = slotName ? slotName.content : `default`
            if (seenSlotNames.has(name)) {
              context.onError(
                createCompilerError(ErrorCodes.X_DUPLICATE_SLOT_NAMES, dirLoc)
              )
              continue
            }
            seenSlotNames.add(name)
          }
          slots.push(
            buildSlot(slotName || `default`, slotProps, children, nodeLoc)
          )
        }
      } else {
        nonSlotChildren.push(child)
      }
    } else {
      nonSlotChildren.push(child)
    }
  }

  if (hasTemplateSlots && nonSlotChildren.length) {
    context.onError(
      createCompilerError(
        ErrorCodes.X_EXTRANEOUS_NON_SLOT_CHILDREN,
        nonSlotChildren[0].loc
      )
    )
  }

  if (!explicitDefaultSlot && !hasTemplateSlots) {
    // implicit default slot.
    slots.push(buildSlot(`default`, undefined, children, loc))
  }

  return createObjectExpression(slots, loc)
}

function buildSlot(
  name: string | ExpressionNode,
  slotProps: ExpressionNode | undefined,
  children: ChildNode[],
  loc: SourceLocation
): Property {
  return createObjectProperty(
    isString(name) ? createSimpleExpression(name, true, loc) : name,
    createFunctionExpression(
      slotProps,
      children,
      children.length ? children[0].loc : loc
    ),
    loc
  )
}

export function buildSlotOutlet(node: ElementNode, context: TransformContext) {
  const { props, children, loc } = node
  const $slots = context.prefixIdentifiers ? `_ctx.$slots` : `$slots`
  let slot: string | CompoundExpressionNode = $slots + `.default`

  // check for <slot name="xxx" OR :name="xxx" />
  let nameIndex: number = -1
  for (let i = 0; i < props.length; i++) {
    const prop = props[i]
    if (prop.type === NodeTypes.ATTRIBUTE) {
      if (prop.name === `name` && prop.value) {
        // static name="xxx"
        const name = prop.value.content
        const accessor = isSimpleIdentifier(name)
          ? `.${name}`
          : `[${JSON.stringify(name)}]`
        slot = `${$slots}${accessor}`
        nameIndex = i
        break
      }
    } else if (prop.name === `bind`) {
      const { arg, exp } = prop
      if (
        arg &&
        exp &&
        arg.type === NodeTypes.SIMPLE_EXPRESSION &&
        arg.isStatic &&
        arg.content === `name`
      ) {
        // dynamic :name="xxx"
        slot = createCompoundExpression(
          [
            $slots + `[`,
            ...(exp.type === NodeTypes.SIMPLE_EXPRESSION
              ? [exp]
              : exp.children),
            `]`
          ],
          loc
        )
        nameIndex = i
        break
      }
    }
  }

  const slotArgs: CallExpression['arguments'] = [slot]
  const propsWithoutName =
    nameIndex > -1
      ? props.slice(0, nameIndex).concat(props.slice(nameIndex + 1))
      : props
  const hasProps = propsWithoutName.length
  if (hasProps) {
    const { props: propsExpression, directives } = buildProps(
      propsWithoutName,
      loc,
      context
    )
    if (directives.length) {
      context.onError(
        createCompilerError(
          ErrorCodes.X_UNEXPECTED_DIRECTIVE_ON_SLOT_OUTLET,
          directives[0].loc
        )
      )
    }
    slotArgs.push(propsExpression)
  }

  if (children.length) {
    if (!hasProps) {
      slotArgs.push(`{}`)
    }
    slotArgs.push(children)
  }

  node.codegenNode = createCallExpression(
    context.helper(RENDER_SLOT),
    slotArgs,
    loc
  )
}
