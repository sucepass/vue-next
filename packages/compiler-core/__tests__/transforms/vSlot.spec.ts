import { CompilerOptions, parse, transform, generate } from '../../src'
import { transformElement } from '../../src/transforms/transformElement'
import { transformOn } from '../../src/transforms/vOn'
import { transformBind } from '../../src/transforms/vBind'
import { transformExpression } from '../../src/transforms/transformExpression'
import { trackSlotScopes } from '../../src/transforms/vSlot'

function parseWithSlots(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)
  transform(ast, {
    nodeTransforms: [
      ...(options.prefixIdentifiers
        ? [transformExpression, trackSlotScopes]
        : []),
      transformElement
    ],
    directiveTransforms: {
      on: transformOn,
      bind: transformBind
    },
    ...options
  })
  return ast
}

describe('compiler: transform component slots', () => {
  test('generate slot', () => {
    const ast = parseWithSlots(
      `
<Comp>
  <Comp v-slot="{ dur }">
    hello {{ dur }}
  </Comp>
</Comp>
`,
      { prefixIdentifiers: true }
    )
    const { code } = generate(ast, { prefixIdentifiers: true })
    console.log(code)
  })
})
