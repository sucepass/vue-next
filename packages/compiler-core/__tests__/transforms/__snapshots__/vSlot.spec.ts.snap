// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: transform component slots dynamically named slots 1`] = `
"const { toString, resolveComponent, createVNode, openBlock, createBlock } = Vue

return function render() {
  const _ctx = this
  const _component_Comp = resolveComponent(\\"Comp\\")
  
  return (openBlock(), createBlock(_component_Comp, null, {
    [_ctx.one]: ({ foo }) => [
      toString(foo),
      toString(_ctx.bar)
    ],
    [_ctx.two]: ({ bar }) => [
      toString(_ctx.foo),
      toString(bar)
    ]
  }, 256 /* DYNAMIC_SLOTS */))
}"
`;

exports[`compiler: transform component slots explicit default slot 1`] = `
"const { toString, resolveComponent, createVNode, openBlock, createBlock } = Vue

return function render() {
  const _ctx = this
  const _component_Comp = resolveComponent(\\"Comp\\")
  
  return (openBlock(), createBlock(_component_Comp, null, {
    default: ({ foo }) => [
      toString(foo),
      toString(_ctx.bar)
    ]
  }))
}"
`;

exports[`compiler: transform component slots implicit default slot 1`] = `
"const { createVNode, resolveComponent, openBlock, createBlock } = Vue

return function render() {
  const _ctx = this
  const _component_Comp = resolveComponent(\\"Comp\\")
  
  return (openBlock(), createBlock(_component_Comp, null, {
    default: () => [
      createVNode(\\"div\\")
    ]
  }))
}"
`;

exports[`compiler: transform component slots named slots 1`] = `
"const { toString, resolveComponent, createVNode, openBlock, createBlock } = Vue

return function render() {
  const _ctx = this
  const _component_Comp = resolveComponent(\\"Comp\\")
  
  return (openBlock(), createBlock(_component_Comp, null, {
    one: ({ foo }) => [
      toString(foo),
      toString(_ctx.bar)
    ],
    two: ({ bar }) => [
      toString(_ctx.foo),
      toString(bar)
    ]
  }))
}"
`;

exports[`compiler: transform component slots nested slots scoping 1`] = `
"const { toString, resolveComponent, createVNode, openBlock, createBlock } = Vue

return function render() {
  const _ctx = this
  const _component_Inner = resolveComponent(\\"Inner\\")
  const _component_Comp = resolveComponent(\\"Comp\\")
  
  return (openBlock(), createBlock(_component_Comp, null, {
    default: ({ foo }) => [
      createVNode(_component_Inner, null, {
        default: ({ bar }) => [
          toString(foo),
          toString(bar),
          toString(_ctx.baz)
        ]
      }),
      toString(foo),
      toString(_ctx.bar),
      toString(_ctx.baz)
    ]
  }))
}"
`;
